File: LeetCode/Apple/TopQuestions/Sorting_Searching/merge.cpp
#include <iostream>
#include <map>
#include <vector>
#include <functional>
#include <algorithm>

using namespace std;

class Solution{
public:
  vector<vector<int> > merge(vector<vector<int> > & intervals)
  {
    //sort the intervals based on the first value
    // sort using a lambda expression
    sort(intervals.begin(), intervals.end(), [](vector<int>& a, vector<int>& b)
    {
            return a[0] > b[0];
    });
    int n;
    vector <vector <int>> merged(n);
    for (int i=0;i<intervals.size();i++)
    {
      if (merged.size()==0|| (merged[-1][1] < intervals[i][0]))
      {
        merged.push_back(intervals[i]);
      }
      else
      {
        merged[-1][1]= max(merged[-1][1],intervals[i][1]);
      }
      return merged;
    }
  }
};

int main()
{
  vector<vector<int> > intervals = {{1,3},{2,6},{8,10},{15,18}};
  Solution output;
  //cout in c++ does not print all of them
  //do elementwise print
  vector <vector<int>> merged;
  merged=output.merge(intervals);
  for (int i=0;i<merged.size();i++)
  {
    for (int j=0;j<merged[i].size();j++)
    {
      cout << merged[i][j] << endl;
    }
  }
  return 0;
}

File: LeetCode/Apple/TopQuestions/Sorting_Searching/merge.py
#Given a collection of intervals, merge all overlappign intervals
#Example 1:
#Input: [[1,3],[2,6],[8,10].[15,18]]
#Output: [[1,6],[8,10],[15,18]]

#Explanation: SInce intervals [1,3] and [2,6] overlap
#merge[1,6]


#Build pairs for comparison,
#Each pair is a List continaing  2 lists
#sort them

class Solution:
    def merge(self, intervals):
        intervals.sort(key=lambda x: x[0])
        print (intervals)
        merged = []
        for interval in intervals:
            # if the list of merged intervals is empty or if the current
            # interval does not overlap with the previous, simply append it.
            if not merged or merged[-1][1] < interval[0]:
                merged.append(interval)
            else:
            # otherwise, there is overlap, so we merge the current and previous
            # intervals.
                merged[-1][1] = max(merged[-1][1], interval[1])
        return merged



#Graph SOlution:
#Connected Components:
#Build a graph between the vectors of
#int condition for merging
#Use Graph to build the solution:
#Build Grpah
class Connected_COmponent:
    def merge(self,intervals):
        return None

if __name__ == "__main__":
    #Datatype is a list of lists
    intervals = [[1,3],[2,6],[8,10],[15,18]]
    output = Solution()
    print ('The new merged interval is:',output.merge(intervals))
